### ISSUE TITLE
reduce_variance gives error in case of RaggedTensor when axis=0

### ISSUE THREAD

**Issue by ashutosh1919** [tensorflow/tensorflow#49606]
As part of #37014 , `reduce_variance` was added for ragged tensors. Although it is working fine for `axis=1`, it gives error for same input if we change to `axis=0`. 

I am attaching the [gist](https://colab.research.google.com/gist/ashutosh1919/b0591ddb485107187b982a08276712be/untitled550.ipynb).

By observing the stack trace, what I think the issue is:
(1) when I explicitly pass input to `ragged_math_ops.reduce_variance`, it gives correct result. So, the problem is not with `reduce_variance` function.
(2) When I call `tf.math.reduce_variance`, it internally call `GlobalDispatcherOp` which iterates over all the dispatchers to see where the op is supported. 
(3) When the op runs for `BinaryRaggedElementwiseDispatcher`, it fails for some reason and that is why execution fails. It didn't reach till the dispatcher associated for `reduce_variance`.
(4) That's why the problem I think is with dispatcher module.

@mihaimaruseac, @edloper - I am not able to exactly find why the error occures in dispatcher. Please help me by pointing to specific direction and will contribute the fix.

**Comment by edloper** [comment]
I don't see any errors (or anything involving ragged tensors or reduce_variance) in the attached gist.  Did you link the right one?

**Comment by edloper** [comment]
I was able to reproduce this problem with:

```
import tensorflow as tf
x = tf.ragged.constant([[1., 2, 3], [4, 5], [6, 7, 8, 9]])
print(tf.math.reduce_variance(x, axis=1))  # succeeds
print(tf.math.reduce_variance(x, axis=0))  # fails
```

The problem arises because the current dispatch mechanism is reactive, not proactive.  In particular, the current dispatch mechanism is a fallback that gets used when the "normal" implementation raises an error.  This design was used to ensure that dispatch didn't add overhead to existing operations, but we have plans to change it to a proactive mechanism, which checks the types of arguments before running the operation.

So the long term fix is to update the dispatch to be proactive, and not a fallback mechanism.

But until we've finished that update, a shorter-term solution is to update reduce_variance to call convert_to_tensor on its input.  Almost all TensorFlow ops call convert_to_tensor on their inputs before processing them.  This ensures that we can pass in non-tensor values, such as numpy arrays or python lists.  The fact that (almost) all TensorFlow ops call convert_to_tensor is what makes the dispatch fallback mechanism work, since this will fail for types such as RaggedTensor.

If we look at some of the other reduce operations in math_ops.py, such as reduce_logsumexp, they do call convert_to_tensor on their input.  I believe that the following change should make dispatch work correctly (though I haven't actually tested it yet):

```
  with ops.name_scope(name):
    input_tensor = ops.convert_to_tensor(input_tensor)       # NEW
    means = reduce_mean(input_tensor, axis=axis, keepdims=True)
```

This change will also have the side benefit that input_tensor won't get converted to a tensor twice.  (In the current code, if input_tensor is a python list or other non-tensor value, then it will get converted twice: once when we call reduce_mean, and again in the expression `input_tensor - means`).

I'd be happy to make that fix if you like; or if you'd prefer to contribute the fix, that would be fine too.
